controllers/clustertopology_controller_compute.go

```diff
@@ -78,8 +84,286 @@ func (r *ClusterTopologyReconciler) getCurrentState(ctx context.Context, cluster
+		if err := setNestedRef(controlPlane, infrastructureMachineTemplate, "spec", "machineTemplate", "infrastructureRef"); err != nil {
+			return nil, errors.Wrap(err, "failed to set the reference to the infrastructure machine template object within the control plane object")
+		}
+	}
+
+	// If it is required to manage the number of replicas for the control plane, set the corresponding field.
+	// NOTE: If the Topology.controlPlane.replicas value is nil, it is assumed that the control plane controller
+	// does not implement support for this field and the ControlPlane object is generated without the number of Replicas.
+	// TODO: uncomment when https://github.com/kubernetes-sigs/cluster-api/issues/4991 is merged
+	// if current.cluster.Spec.Topology.ControlPlane.Replicas != nil {
+	if err := unstructured.SetNestedField(controlPlane.UnstructuredContent(), int64(current.cluster.Spec.Topology.ControlPlane.Replicas), "spec", "replicas"); err != nil {
+		return nil, errors.Wrap(err, "failed to set the replicas in the control plane object")
+	}
+	// }
+
+	// Sets the desired Kubernetes version for the control plane.
+	// NOTE: Supporting this field is a requirement for control plane providers being compatible with ClusterClass.
+	// TODO: improve this logic by adding support for version upgrade component by component
+	if err := unstructured.SetNestedField(controlPlane.UnstructuredContent(), current.cluster.Spec.Topology.Version, "spec", "version"); err != nil {
+		return nil, errors.Wrap(err, "failed to set the version in the control plane object")
+	}
+
+	return controlPlane, nil
+}
+
+// computeCluster computes the desired state for the Cluster object.
+// NOTE: Some fields of the Cluster’s fields contribute to defining how a Cluster should look like (e.g. Cluster.Spec.Topology),
+// while some other fields should be managed as part of the actual cluster (e.g. Cluster.Spec.ControlPlaneRef); in this func
+// we are concerned only about the latest group of fields.
+func computeCluster(current *clusterTopologyState, infrastructureCluster, controlPlane *unstructured.Unstructured) *clusterv1.Cluster {
+	cluster := &clusterv1.Cluster{}
+	current.cluster.DeepCopyInto(cluster)
+
+	// Enforce the topology labels.
+	// NOTE: The cluster label is added at creation time so this object could be read by the ClusterTopology
+	// controller immediately after creation, even before other controllers are going to add the label (if missing).
+	if cluster.Labels == nil {
+		cluster.Labels = map[string]string{}
+	}
+	cluster.Labels[clusterv1.ClusterLabelName] = cluster.Name
+	cluster.Labels[clusterv1.ClusterTopologyLabelName] = ""
+
+	// Set the references to the infrastructureCluster and controlPlane objects.
+	// NOTE: Once set for the first time, the references are not expected to change.
+	cluster.Spec.InfrastructureRef = objToRef(infrastructureCluster)
+	cluster.Spec.ControlPlaneRef = objToRef(controlPlane)
+
+	return cluster
+}
+
+type templateToInput struct {
```

Comment:
```suggestion
type templateToObjectInput struct {
```
nit (if you want, I think that was our pattern in other cases)

Reply:
This input struct is used by two func, also it is internal